ghga_service_chassis_lib/pubsub.py

```diff
@@ -0,0 +1,205 @@
+        return True
+    except jsonschema.exceptions.ValidationError as exc:
+        logging.error(
+            "%s: Message payload does not comform to JSON schema.",
+            datetime.now().isoformat(timespec="milliseconds"),
+        )
+        logging.exception(exc)
+        if raise_on_exception:
+            raise exc
+        return False
+
+
+def callback_wrapper_factory(
+    exec_on_message: Callable, json_schema: Optional[dict] = None
+) -> Callable:
+    """Generates a callback function that is executed whenever a message reaches
+    the queue. It performs logging, message validation against a json schema (if provided)
+    and, finally, executes the function `exec_on_message`."""
+
+    def callback(
+        channel: pika.channel.Channel,
+        method: pika.spec.Basic.Deliver,
+        _: pika.spec.BasicProperties,
+        body: str,
+    ):
+        """A wrapper around the actual function that is executed
+        once a message arrives:"""
+
+        logging.info(
+            " [x] %s: Message received",
+            datetime.now().isoformat(timespec="milliseconds"),
+        )
+
+        message = json.loads(body)
+
+        if json_schema:
+            valid = validate_message(message, json_schema)
+            if not valid:
+                channel.basic_nack(delivery_tag=method.delivery_tag, requeue=False)
+
+        try:
+            exec_on_message(message)
+        except (MaxAttemptsReached, ValueError):
+            channel.basic_nack(delivery_tag=method.delivery_tag, requeue=False)
+        else:
+            channel.basic_ack(delivery_tag=method.delivery_tag)
+
+    return callback
+
+
+@dataclass
+class AmqpTopic:
+    """A base class to connect and iteract to/with a RabbitMQ host
+    via the `topic` exchange type.
+
+    Args:
+        connection_params [Type(pika.connection.Parameters)]:
+            An object of class pika.connection.Parameters that configures
+            the connection to the RabbitMQ broker.
+        topic_name (str):
+            The name of the topic (only use letters, numbers, and "_").
+            The queue binding key as well as the names for the associated exchange
+            and queue will be derived from this string.
+        service_name (str):
+            A name that uniquely identifies the service. This will be used to
+            ensure that messages will not be duplicates between instances of
+            the same service.
+        json_schema (Optional[dict]):
+            Optional. If provided, the message body will be validated against this
+            json schema.
+
+    Naming patterns for Exchanges and Queues:
+        Exchanges will always be named according to the `topic_name`.
+        Queue names will be generated by concatenating the `service_name`
+        and the `topic_name`
+    """
+
+    connection_params: Type[pika.connection.Parameters]
+    topic_name: str
+    service_name: str
+    json_schema: Optional[dict] = None
```

Comment:
Can we have a default constructor that takes these members as arguments and returns the AmqpTopic Object? examples/publisher.py line no 37 makes use of this concept.

Reply:
Since this is a dataclass, it creates a constructor for us implicitly. 

Of course, we could use a default python class with an "__init__" function, but that init function would just forward the function params to self. So this will just be more elaborate and we lose on the type annotation feature of dataclasses.

That's why I used a dataclass here. Do you agree?

Reply:
Yes, I agree.. The dataclass has an implicit default constructor.